### ISSUE TITLE
1.21.1 Crash - Crash when changing dimensions

### ISSUE THREAD

**Issue by VOrlando520** [DrexHD/MagicParticles#5]
Crashed the server randomly with the following crash report
https://pastebin.com/hfWxcA6P

**Comment by DrexHD** [comment]
That's a weird crash. Iterating the player list on the server thread should not cause any issues in a regular environment. Can you sent a list of your mods and if possible how to reproduce this (did someone leave / join / change world just before the crash)

**Comment by VOrlando520** [comment]
Probably an optimization mod then from my guess. Idk if there is a command on Fabric to output the list of mods on a server, but it was a very random crash that is hard to pinpoint with 50-60 people on a server all doing random things.

**Comment by DrexHD** [comment]
Can you sent the full crash log (it has the mod list included and some additional information that might help pinpoint the problem). You can send it to me in DMs via discord (my support discord is linked in my profile) if you don't want to leak anything here (like player names + locations and mod list)

**Comment by VOrlando520** [comment]
> Can you sent the full crash log (it has the mod list included and some additional information that might help pinpoint the problem). You can send it to me in DMs via discord (my support discord is linked in my profile) if you don't want to leak anything here (like player names + locations and mod list)

https://pastebin.com/LGjh2a2i

**Comment by DrexHD** [comment]
I don't see any horribly bad mods in that list at a quick glance. I was however hoping for the **A detailed walkthrough of the error, its code path and all known details is as follows:** section, because I want to check if any other thread stack trace could be hinting at something modifying the player list.

**Comment by VOrlando520** [comment]
> I don't see any horribly bad mods in that list at a quick glance. I was however hoping for the **A detailed walkthrough of the error, its code path and all known details is as follows:** section, because I want to check if any other thread stack trace could be hinting at something modifying the player list.

Ah, was hoping you'd of seen something I didnt. Sadly I had gotten this crash 3 more times and had to disable it until we can find the underlying cause. That was the full crash report though. Since it has been disabled no crashes though. So im lost :\

**Comment by DrexHD** [comment]
Looks like the Thread Dump I was looking for is only added to the crash report when the crash is caused by Server Watchdog. I don't think I can do anything about this crash at the moment unfortunately. I would need some way to (at least somewhat) consistently reproduce the issue to start debugging it! With such a big mod list and it presumably only happening very rarely and no further info I don't really have a chance :/

**Comment by VOrlando520** [comment]
> Looks like the Thread Dump I was looking for is only added to the crash report when the crash is caused by Server Watchdog. I don't think I can do anything about this crash at the moment unfortunately. I would need some way to (at least somewhat) consistently reproduce the issue to start debugging it! With such a big mod list and it presumably only happening very rarely and no further info I don't really have a chance :/

Im going to flip server watchdog back on and reenable this for my next restart, see if i can get this to happen again for a more detailed stacktrace for ya

**Comment by DrexHD** [comment]
[Please try it with the referenced commit](https://nightly.link/DrexHD/MagicParticles/actions/runs/9752498199), it should prevent the crashes and create a crash report (without shutting down) with the desired information instead. I am skeptical it will actually reveal the cause, but it is my best shot and should at least prevent the crashes for you! If you use this version you should be able to find a crash report with magic-particles in the name (if the issue occurs again) that you should share here!

**Comment by VOrlando520** [comment]
> [Please try it with the referenced commit](https://nightly.link/DrexHD/MagicParticles/actions/runs/9752498199), it should prevent the crashes and create a crash report (without shutting down) with the desired information instead. I am skeptical it will actually reveal the cause, but it is my best shot and should at least prevent the crashes for you! If you use this version you should be able to find a crash report with magic-particles in the name (if the issue occurs again) that you should share here!

Will report back with my findings when it happens!

**Comment by VOrlando520** [comment]
> [Please try it with the referenced commit](https://nightly.link/DrexHD/MagicParticles/actions/runs/9752498199), it should prevent the crashes and create a crash report (without shutting down) with the desired information instead. I am skeptical it will actually reveal the cause, but it is my best shot and should at least prevent the crashes for you! If you use this version you should be able to find a crash report with magic-particles in the name (if the issue occurs again) that you should share here!

[crash-2024-07-02_02.01.23-magic-particles.txt](https://github.com/user-attachments/files/16060735/crash-2024-07-02_02.01.23-magic-particles.txt)
unsure if that will help you but here ya go

**Comment by DrexHD** [comment]
Unfortunately that didn't help :/

**Comment by DrexHD** [comment]
_Whoops didn't meant to close and send that._
Unfortunately that didn't help :/ I am not really sure how else I could investigate this (and also why this doesn't happen with any other mod). By the way you seem to have some mod spam creating tons of Timer threads, which definitely isn't great (and could be related :shrug:).

**Comment by VOrlando520** [comment]
> _Whoops didn't meant to close and send that._ Unfortunately that didn't help :/ I am not really sure how else I could investigate this (and also why this doesn't happen with any other mod). By the way you seem to have some mod spam creating tons of Timer threads, which definitely isn't great (and could be related 🤷).

I had noticed that as well, but couldn't narrow down what it could be.

**Comment by VOrlando520** [comment]
So i get this going now before everyone times out
https://pastebin.com/YiEepVcx

Unsure if that will help you out.

**Comment by DrexHD** [comment]
I have added something else to try and figure out wtf is happening here. Can you please try with the latest commit and (if it happens again, aka. custom crash report) send the log file.

**Comment by VOrlando520** [comment]
[crash-2024-07-23_19.47.32-magic-particles.txt](https://github.com/user-attachments/files/16362213/crash-2024-07-23_19.47.32-magic-particles.txt)

Yeah doesnt look like this is helping unveil anything. Sadly just going to have to remove it because it only times out with this mod installed and going.

**Comment by DrexHD** [comment]
You got a "crash report" with the most recent version? Can you please share (part of) the server log. There _should_ be a line saying something like `Potential cause: [...]` followed by some more info.

**Comment by VOrlando520** [comment]
Ah could of sworn that was the latest version from the commits artifacts. Ill keep an eye out, see what it does next.

**Comment by DrexHD** [comment]
It's very much possible you already got the latest commit artifacts (I can't tell from crash report). I was just saying, **if** that crash occurred while you had the latest artifact ([3 days ago](https://github.com/DrexHD/MagicParticles/actions/runs/10030270111)), I need to see your server logs (latest.log) at the timestamp of this happening, because there should be some logs printing (hopefully) further useful information on the issue!

**Comment by VOrlando520** [comment]
> It's very much possible you already got the latest commit artifacts (I can't tell from crash report). I was just saying, **if** that crash occurred while you had the latest artifact ([3 days ago](https://github.com/DrexHD/MagicParticles/actions/runs/10030270111)), I need to see your server logs (latest.log) at the timestamp of this happening, because there should be some logs printing (hopefully) further useful information on the issue!

so nothing crashes with the last two artifacts, but ever since we added this back it just times people out after awhile

**Comment by DrexHD** [comment]
And there is also no custom crash reports anymore?

**Comment by VOrlando520** [comment]
> And there is also no custom crash reports anymore?

Nope, just times out, then after awhile we have to restart to fix the issue.

**Comment by DrexHD** [comment]
Is there anything special in the logs during (at the time of) the timeouts?

**Comment by VOrlando520** [comment]
Only thing that stands out is a bunch of luckperms timeouts for mariadb

**Comment by VOrlando520** [comment]
Alright, I am back and on 1.21.1 and can confirm a crash does still happen, but it is only when switching dimensions. Basically, have the particles on, switch to a dimenion and it will crash the entire server (it isnt consistent, but has happened a handful of times in the last two weeks)

**Comment by VOrlando520** [comment]
Another crash report, another CME, i do not get it https://pastebin.com/Pg5qVsVU
